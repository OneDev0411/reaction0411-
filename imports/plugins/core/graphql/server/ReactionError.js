import ExtendableError from "es6-error";

export default class ReactionError extends ExtendableError {
  constructor(error, message, eventData = {}) {
    super(message);
    this.eventData = eventData;
    this.error = error;
  }
}
