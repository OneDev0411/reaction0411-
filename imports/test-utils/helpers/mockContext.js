const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getAbsoluteUrl: jest.fn().mockName("getAbsoluteUrl").mockImplementation((path) => {
    const adjustedPath = path[0] === "/" ? path : `/${path}`;
    return `https://app.mock${adjustedPath}`;
  }),
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  mutations: {},
  queries: {},
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

/**
 * @summary Returns a mock collection instance with the given name
 * @param {String} collectionName The collection name
 * @returns {Object} Mock collection instance
 */
export function mockCollection(collectionName) {
  return {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    bulkWrite: jest.fn().mockName(`${collectionName}.bulkWrite`).mockReturnValue(Promise.resolve({
      nMatched: 2,
      nModified: 2,
      result: {
        writeErrors: []
      }
    })),
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    findOneAndDelete: jest.fn().mockName(`${collectionName}.findOneAndDelete`),
    findOneAndUpdate: jest.fn().mockName(`${collectionName}.findOneAndUpdate`),
    fetch: jest.fn().mockName(`${collectionName}.fetch`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
}

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "ExampleIOUPaymentRefunds",
  "Groups",
  "MediaRecords",
  "NavigationItems",
  "NavigationTrees",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "SimpleInventory",
  "Tags",
  "Templates",
  "Themes",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = mockCollection(collectionName);
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
