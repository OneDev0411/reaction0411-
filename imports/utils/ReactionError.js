import ExtendableError from "es6-error";

/**
 * @name ReactionError
 * @class
 * @memberof GraphQL
 * @summary A type of error that allows additional details to be provided, which can
 *   then be sent back to the client for a GraphQL request.
 */
export default class ReactionError extends ExtendableError {
  constructor(error, message, eventData = {}) {
    super(message);
    this.eventData = eventData;
    this.error = error;
    // Newer versions of DDP use this property to signify that an error
    // can be sent back and reconstructed on the calling client.
    this.isClientSafe = true;
    // DDP expects this to be in a property named `reason`
    this.reason = message;
    // DDP expects this to be in a property named `details`
    this.details = eventData;
  }

  clone() {
    return new ReactionError(this.error, this.reason, this.eventData);
  }
}
