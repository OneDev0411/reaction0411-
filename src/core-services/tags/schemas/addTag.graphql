"Input for `addTag` mutation"
input AddTagInput {
  "An optional string identifying the mutation call, which will be returned in the response payload"
  clientMutationId: String

  "Title to display to customers"
  displayTitle: String

  "Hero media URL"
  heroMediaUrl: String

  "Whether the tag is visible"
  isVisible: Boolean!

  "Tag metafields"
  metafields: [MetafieldInput]

  "Unique name of the tag"
  name: String!

  "The shop that owns the tag"
  shopId: ID!

  "The tag slug. If left blank, the name will be slugified and saved as the slug"
  slug: String
}

"Response payload for `addTag` mutation"
type AddTagPayload {
  "The same string you sent with the mutation params, for matching mutation calls with their responses"
  clientMutationId: String

  "The shop that owns the tag"
  shopId: ID!

  "The newly-created tag"
  tag: Tag!
}

extend type Mutation {
  "Adds a new tag"
  addTag(
    "Mutation input"
    input: AddTagInput!
  ): AddTagPayload!
}
