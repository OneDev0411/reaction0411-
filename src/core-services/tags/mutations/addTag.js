import Random from "@reactioncommerce/random";
import ReactionError from "@reactioncommerce/reaction-error";
import getSlug from "@reactioncommerce/api-utils/getSlug.js";
import { Tag as TagSchema } from "../simpleSchemas.js"; // TODO: update schemas

/**
 * @name Mutation.addTag
 * @method
 * @memberof Routes/GraphQL
 * @summary Add a tag
 * @param {Object} context -  an object containing the per-request state
 * @param {Object} input - mutation input
 * @returns {Promise<Object>} AddTagPayload
 */
export default async function addTag(context, input) {
  // Check for owner or admin permissions from the user before allowing the mutation
  const { shopId, name, isVisible, displayTitle, metafields, heroMediaUrl, slug: slugInput } = input;
  const { appEvents, checkPermissions, collections } = context;
  const { Tags } = collections;

  await checkPermissions(["admin", "owner"], shopId);

  let slug = name;
  if (typeof slugInput === "string" && slugInput.trim().length > 0) {
    slug = slugInput;
  }

  const now = new Date();
  const tag = {
    _id: Random.id(),
    isDeleted: false,
    isTopLevel: false,
    isVisible,
    slug: getSlug(slug),
    metafields,
    name,
    displayTitle,
    heroMediaUrl,
    shopId,
    createdAt: now,
    updatedAt: now
  };

  TagSchema.validate(tag);

  try {
    const { result } = await Tags.insertOne(tag);

    if (result.ok !== 1) {
      throw new ReactionError("server-error", "Unable to create tag");
    }

    await appEvents.emit("afterTagCreate", tag);

    return tag;
  } catch ({ message }) {
    // Mongo duplicate key error.
    if (message.includes("E11000") && message.includes("slug")) {
      throw new ReactionError("error", `Slug ${tag.slug} is already in use`);
    }

    throw new ReactionError("error", message);
  }
}
