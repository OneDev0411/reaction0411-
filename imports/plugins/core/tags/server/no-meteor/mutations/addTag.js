import Random from "@reactioncommerce/random";
import ReactionError from "@reactioncommerce/reaction-error";
import { Tag as TagSchema } from "/imports/collections/schemas";

/**
 * @name Mutation.addTag
 * @method
 * @memberof Routes/GraphQL
 * @summary Add a tag
 * @param {Object} context -  an object containing the per-request state
 * @param {Object} input - mutation input
 * @return {Promise<Object>} AddTagPayload
 */
export default async function addTag(context, input) {
  // Check for owner or admin permissions from the user before allowing the mutation
  const { shopId } = input;
  const { appEvents, collections, userHasPermission } = context;
  const { Tags } = collections;

  if (!userHasPermission(["owner", "admin"], shopId)) {
    throw new ReactionError("access-denied", "User does not have permission");
  }

  const { clientMutationId } = input;

  const now = new Date();
  const tag = {
    _id: Random.id(),
    isVisible: input.isVisible,
    name: input.name,
    displayTitle: input.displayTitle,
    createdAt: now,
    updatedAt: now
  };

  TagSchema.validate(tag);
  const { result } = await Tags.insertOne(tag);

  if (result.ok !== 1) {
    throw new ReactionError("server-error", "Unable to create tag");
  }

  await appEvents.emit("afterTagCreate", tag);

  return {
    clientMutationId,
    tag
  };
}
