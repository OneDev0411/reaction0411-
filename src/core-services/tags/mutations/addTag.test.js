import mockContext from "@reactioncommerce/api-utils/tests/mockContext.js";
import ReactionError from "@reactioncommerce/reaction-error";
import addTag from "./addTag.js";

beforeEach(() => {
  jest.resetAllMocks();
});

test("calls mutations.addTag and returns the AddTagPayload on success", async () => {
  mockContext.checkPermissionsLegacy.mockReturnValueOnce(Promise.resolve(null));
  mockContext.collections.Tags.insertOne.mockReturnValueOnce({ result: { ok: 1 } });

  const input = {
    shopId: "1234",
    name: "shirt",
    displayTitle: "Shirt",
    isVisible: true
  };
  const result = await addTag(mockContext, input);

  expect(result).toBeDefined();
  expect(mockContext.collections.Tags.insertOne).toHaveBeenCalled();
});

test("calls mutations.addTag and throws for non admins", async () => {
  mockContext.checkPermissionsLegacy.mockImplementation(() => {
    throw new ReactionError("access-denied", "Access Denied");
  });
  mockContext.collections.Tags.insertOne.mockReturnValueOnce({ result: { ok: 1 } });

  const result = addTag(mockContext, {});
  expect(result).rejects.toThrowErrorMatchingSnapshot();
  expect(mockContext.collections.Tags.insertOne).not.toHaveBeenCalled();
});
