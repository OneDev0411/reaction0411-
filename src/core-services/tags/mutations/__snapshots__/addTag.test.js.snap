// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls mutations.addTag and throws for non admins 1`] = `"Access Denied"`;
